test/integration/job/job_test.go

```diff
@@ -51,13 +51,332 @@ import (
+						OnExitCodes: &batchv1.PodFailurePolicyOnExitCodesRequirement{
+							Operator: batchv1.PodFailurePolicyOnExitCodesOpIn,
+							Values:   []int32{5, 6, 7},
+						},
+					},
+				},
+			},
+		},
+	}
+	podStatusMatchingOnExitCodesTerminateRule := v1.PodStatus{
+		Phase: v1.PodFailed,
+		ContainerStatuses: []v1.ContainerStatus{
+			{
+				Name: "main-container",
+				State: v1.ContainerState{
+					Terminated: &v1.ContainerStateTerminated{
+						ExitCode: 5,
+					},
+				},
+			},
+		},
+	}
+	wFinalizers := true
+	defer featuregatetesting.SetFeatureGateDuringTest(t, feature.DefaultFeatureGate, features.JobTrackingWithFinalizers, wFinalizers)()
+	defer featuregatetesting.SetFeatureGateDuringTest(t, feature.DefaultFeatureGate, features.JobPodFailurePolicy, true)()
+	for _, policy := range []metav1.DeletionPropagation{metav1.DeletePropagationOrphan} {
+		t.Run(string(policy), func(t *testing.T) {
+			closeFn, restConfig, clientSet, ns := setup(t, "simple")
+			defer closeFn()
+			informerSet := informers.NewSharedInformerFactory(clientset.NewForConfigOrDie(restclient.AddUserAgent(restConfig, "controller-informers")), 0)
+			// Make the job controller significantly slower to trigger race condition.
+			restConfig.QPS = 1
+			restConfig.Burst = 1
+			jc, ctx, cancel := createJobControllerWithSharedInformers(restConfig, informerSet)
+			defer func() {
+				cancel()
+			}()
+			restConfig.QPS = 200
+			restConfig.Burst = 200
+			podgc := podgc.NewPodGCInternal(ctx,
+				clientSet,
+				informerSet.Core().V1().Pods(),
+				informerSet.Core().V1().Nodes(),
+				1,
+				100*time.Millisecond,
+				time.Second)
+			informerSet.Start(ctx.Done())
+			go jc.Run(ctx, 1)
+			//	go podgc.Run(ctx)
+
+			jobObj, err := createJobWithDefaults(ctx, clientSet, ns.Name, &job)
+			if err != nil {
+				t.Fatalf("Failed to create Job: %v", err)
+			}
+			// if !hasJobTrackingAnnotation(jobObj) {
+			// 	t.Error("apiserver didn't add the tracking annotation")
+			// }
+			validateJobPodsStatus(ctx, t, clientSet, jobObj, podsByStatus{
+				Active: 5,
+				Ready:  pointer.Int32(0),
+			}, wFinalizers)
+			pods, err := clientSet.CoreV1().Pods(jobObj.Namespace).List(ctx, metav1.ListOptions{})
+			if err != nil {
+				t.Fatalf("Failed to list Job Pods: %v", err)
+			}
+			t.Logf("Got pods for job: %v", len(pods.Items))
+			failedIndex := 1
+			wg := sync.WaitGroup{}
+			wg.Add(1)
+			go func() {
+				t.Logf("__________________ ____________________ Inside the routine ___________________ ")
+				err := wait.PollImmediate(500*time.Millisecond, time.Minute, func() (bool, error) {
+					failedPodUpdated, err := clientSet.CoreV1().Pods(jobObj.Namespace).Get(ctx, pods.Items[failedIndex].Name, metav1.GetOptions{})
+					if err != nil {
+						t.Logf("Failed to get the failed pod: %v", err)
+						return true, err
+					}
+					if len(failedPodUpdated.Finalizers) == 0 {
+						t.Logf("================ YEAH, finalizer removed!!!!!")
```

Comment:
Debugging leftovers? Quite a few :wink:

Reply:
Yeah, now cleaned up and with comments :)